ambari-server/src/main/java/org/apache/ambari/server/topology/addservice/AddServiceInfo.java

```diff
@@ -38,18 +39,32 @@
   private final Map<String, Map<String, Set<String>>> newServices;
   private final RequestStageContainer stages;
   private final Configuration config;
-
-  public AddServiceInfo(AddServiceRequest request, String clusterName, Stack stack, Configuration config, RequestStageContainer stages, Map<String, Map<String, Set<String>>> newServices) {
+  private final Optional<LayoutRecommendationInfo> recommendationInfo;
+
+  public AddServiceInfo(AddServiceRequest request,
+                        String clusterName,
+                        Stack stack,
+                        Configuration config,
+                        RequestStageContainer stages,
+                        Map<String, Map<String,
+                        Set<String>>> newServices,
+                        Optional<LayoutRecommendationInfo> recommendationInfo) {
     this.request = request;
     this.clusterName = clusterName;
     this.stack = stack;
     this.newServices = newServices;
     this.stages = stages;
     this.config = config;
+    this.recommendationInfo = null != recommendationInfo ? recommendationInfo : Optional.empty();
```

Comment:
I think the field and constructor parameter should be of type `LayoutRecommendationInfo`, not `Optional< LayoutRecommendationInfo>`, and `getRecommendationInfo()` should wrap it in `Optional.ofNullable()`.  Makes the constructor a bit simpler, and gets rid of warnings in IDEA.